src/FSharpVSPowerTools.Core/Utils.fs

```diff
@@ -27,60 +48,136 @@ module List =
             | _ -> List.rev acc
         loop [] xs
 
+    /// Fold over the list passing the index and element at that index to a folding function
+    let foldi (folder:'State -> int -> 'T -> 'State) (state:'State) (xs:'T list) =        
+        match xs with 
+        | [] -> state
+        | _ -> 
+            let f = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt folder
+            let rec loop idx s xs = 
+                match xs with 
+                | [] -> s
+                | h::t -> loop (idx+1) (f.Invoke(s,idx,h)) t
+            loop 0 state xs
+
+
+    let inline groupBy (keyfn:'T->'Key) (list: 'T list) =
+        let dict = Dictionary<'Key,ResizeArray<'T>> (HashIdentity.Structural)
+        // Build the groupings
+        let rec loop list =
+            match list with
+            | v :: t -> 
+                let key = keyfn v
+                let ok,prev = dict.TryGetValue key
+                if ok then
+                    prev.Add v
+                else 
+                    let prev = new ResizeArray<'T> 1
+                    dict.[key] <- prev
+                    prev.Add v
+                loop t
+            | _ -> ()
+        loop list
+        dict  // Return the list-of-lists.
+        |> Seq.map (fun group -> (group.Key, group.Value.ToArray() |> List.ofArray))
+        |> Seq.toList
+
+
 [<RequireQualifiedAccess>]
-[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
+[<CompilationRepresentation (CompilationRepresentationFlags.ModuleSuffix)>]
 module Array =
     let inline private checkNonNull argName arg = 
         match box arg with 
         | null -> nullArg argName 
         | _ -> ()
 
+    /// Optimized arrays equality. ~100x faster than `array1 = array2` on strings.
+    /// ~2x faster for floats
+    /// ~0.8x slower for ints
+    let inline areEqual (x: 'a[]) (y: 'a[]) =
+        match x, y with
+        | null, null -> true
+        | [||], [||] -> true
+        | null, _ | _, null -> false
+        | _ when x.Length <> y.Length -> false
+        | _ ->
+            let mutable break' = false
+            let mutable i = 0
+            let mutable result = true
+            while i < x.Length && not break' do
+                if x.[i] <> y.[i] then 
+                    break' <- true
+                    result <- false
+                &i += 1
+            result
+
+
     /// Returns true if one array has another as its subset from index 0.
-    let startsWith (prefix: _ []) (whole: _ []) =
-        let rec loop index =
-            if index > prefix.Length - 1 then true
-            elif index > whole.Length - 1 then false
-            elif prefix.[index] = whole.[index] then loop (index + 1)
-            else false
-        prefix.Length = 0 || loop 0
+    let startsWith (prefix: _ []) (whole: _ []) =        
```

Comment:
```
> original startswith 0.02ms
startswithIf 0.02ms
Ran 100 times
Ratio:  1.0
val it : unit = ()
>
original startswith 0.15ms
startswithIf 0.1ms
Ran 100 times
Ratio:  1.5
val it : unit = ()
> original startswith 20.36ms
startswithIf 11.89ms
Ran 100 times
Ratio:  1.712363331
val it : unit = ()
>
startswithMatch 0.06ms
startswithIf 0.04ms
Ran 100 times
Ratio:  1.5
val it : unit = ()
> startswithMatch 0.15ms
startswithIf 0.07ms
Ran 100 times
Ratio:  2.142857143
val it : unit = ()
> startswithMatch 11.92ms
startswithIf 12.38ms
Ran 100 times
Ratio:  0.9628432956
val it : unit = ()
```
test code compiled in release mode https://gist.github.com/cloudRoutine/88143149f02f766b1389